### DISCUSSION TITLE
Mobile Translation Pipeline

### DISCUSSION THREAD

**Discussion by kgellci** [MobileNativeFoundation/discussions#10] [General]
Hey all! 

Would love to learn about how others have setup their mobile translation pipeline. I can give some detail about how we have it setup at Lyft.

For both iOS and Android, our string extraction and distribution is fully automated and works based on our release schedule (Weekly release branch cuts from latest master. Release branches bake in Beta for a week. Slow rollout to App Store and Play Store). 

**English String Extraction**

We have tooling which bundles all of the native English string files for iOS and Android (Localizable.strings, Strings.xml, etc.) and uploads them to some S3 bucket. Each platform uploads its own bundle, Android and iOS string files are not mixed into a single bundle. The bundle contains the directory paths with the nested string files (imagine taking your repo and deleting every file that is not a supported strings file).

We have an internal translation pipeline web service which has several cron jobs. On a daily basis, it picks up the latest master branch English string bundles for both platforms and uploads the files to our 3rd party translation management service (Smartling). There is another daily job which downloads all of the translated files from Smartling for a given platform and opens a PR on master with the diff of translated strings (The PR is auto merged once tests pass).

**Translation Distribution**

With the above system, we give a guarantee to teams that their features will be auto translated, but there is still the possibility that a team merges a feature into master right before the release branch is cut. We have one more job which does a final string distribution to the release branch ~ 2 days after the release branch is cut.  We additionally have tooling / alarms which can inform us if we have release branches which are missing translated strings.

**Extra**

One additional note for iOS, we currently support Storyboard and Xib localization. This meaning that strings do not need to be placed in Localizable.strings files, this is actually natively supported by Apple tooling (ibtool). Since we initially wanted to do this extraction on some non-mac hardware, we could not actually use ibtool to translate the storyboard and xib strings so we wrote our own tool in Python which essentially does what ibtool does in regards to extracting English strings for localization (with some extra Lyft specific customization).

**Comment by AfrazCodes** [comment]
Hey Kris,

Thanks for sharing this, super insightful. Seems like there are some similarities (automation, extraction, distribution) with how we are localizing strings for Outlook iOS.

### [LocalizedStringKit](https://github.com/microsoft/LocalizedStringKit)

We leverage a native library and related python generation package which automates a great deal of our process - further detailed below.

### Checking In English Strings

Developers are free to add whatever strings they need directly to the project so long as its wrapped in the native frameworks API.
```swift
Localized("foo", "comment")
```
Under the hood, this is just resolving a localized string from a given table, bundle, and optional key. 

### Generating String Tables

Before checking in, one can run a `localize` command through CLI which executes the python package. This package scans all the sources files in the project and generates English string tables with appropriate keys.

### Getting Actual Translations

We then have a cron job/recurring pipeline which sends our English string table to an internal service for translation. Response of the pipeline includes string tables for other languages. That pipeline thereafter commits it directly to the project.

## Multiple Bundles (Watch Extension, other targets)

We also support specifying multiple bundles to our pipeline job for translation to more easily allow decoupling of targets and their respective bundles. This supports things like a watchOS Extension, iOS 14 Widget, etc.

## Syncing Strings

More recently, we've been working on a process to offload all of our strings to a Content Delivery Network (CDN) and sync localized language strings to the end user. The goal of this project is to reduce the app size we ship to the App Store by excluding the majority of strings (we still package some of course :) ).

You can read more about [String Offloading & Sync here](https://medium.com/microsoft-mobile-engineering/app-thinning-syncing-localized-strings-to-outlook-for-ios-98ec2479e9d6).

**Comment by ghost** [comment]
@kgellci How do you keep iOS and Android strings consistent?

**Reply by kgellci** [reply]
We do not enforce consistency of strings between the two platforms. It is up to the individual feature developers to ensure that the English strings for their teams features between the two platforms are consistent.

**Reply by keith** [reply]
This is an interesting idea though. I guess do we always expect things to be equivalent? Maybe we should actually be trying to unify that way

**Reply by ghost** [reply]
An alternative is for strings to be agreed for each feature and added to the localisation service (possibly role given to PM).  Then automatic generation of strings code back to each platform's repo.  PRs can be automatically raised with the new/changed strings.  Then translations go through same automated route when they are ready.

**Comment by changusmc** [comment]
We use largely the same process as @kgellci at Dropbox.

## English String sync
We have job that runs every day to push all English strings to our translation service (Smartling) from our main branch.

We also download any strings from our translation service and commit those strings into the repo and run a CI build and lint job to validate those strings didn't break the build. If they broke the build, we auto-revert the translation and auto-open a JIRA ticket that blocks the release from going out.

Two days before we release the app, we sync the strings to the RC branch and run the same validations.

We have a guarantee for our developers that if they get their strings in before the branch cut, it will make it into the stable release. If they cherry-pick into the stable branch, then they need to ensure the strings made it in.

**Comment by afrishberg** [comment]
How do you check for UI not breaking in other languages? I know German and Russian are the usual suspects.

**Reply by ghost** [reply]
Fastlane Snapshot can be used for checking multiple screens across different languages and device sizes (https://docs.fastlane.tools/actions/snapshot/), so it's not just for App Store screenshots, but for testing too.
